jetty-webapp/src/main/java/org/eclipse/jetty/webapp/DecoratingListener.java

```diff
@@ -0,0 +1,180 @@
+ * be added to the {@link WebAppContext#getObjectFactory()} as a
+ * {@link Decorator}.
+ */
+public class DecoratingListener implements ServletContextAttributeListener
+{
+    public static final String DECORATOR_ATTRIBUTE = "org.eclipse.jetty.webapp.decorator";
+    private static final Logger LOG = Log.getLogger(DecoratingListener.class);
+    private static final MethodType decorateType;
+    private static final MethodType destroyType;
+
+    static
+    {
+        try
+        {
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+            decorateType = MethodType.methodType(Object.class, Object.class);
+            destroyType = MethodType.methodType(Void.TYPE, Object.class);
+            // Ensure we have a match
+            lookup.findVirtual(Decorator.class, "decorate", decorateType);
+            lookup.findVirtual(Decorator.class, "destroy", destroyType);
+        }
+        catch (Exception e)
+        {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    private final WebAppContext _context;
+    private final String _attributeName;
+    private Decorator _decorator;
+
+    public DecoratingListener()
+    {
+        this(null, null);
+    }
+
+    public DecoratingListener(String attributeName)
+    {
+        this(null, attributeName);
+    }
+
+    public DecoratingListener(WebAppContext context)
+    {
+        this(context, null);
+    }
+
+    public DecoratingListener(WebAppContext context, String attributeName)
+    {
+        _context = context == null ? WebAppContext.getCurrentWebAppContext() : context;
+        _attributeName = attributeName == null ? DECORATOR_ATTRIBUTE : attributeName;
+        if (_context.getAttribute(getClass().getName()) == null)
+            _context.setAttribute(getClass().getName(), attributeName);
+        else if (LOG.isDebugEnabled())
+            LOG.debug("DecoratingListener already set for {} in {}",
+                _context.getAttribute(getClass().getName()), _context);
+        Object decorator = _context.getAttribute(_attributeName);
+        if (decorator != null)
+        {
+            _decorator = asDecorator(decorator);
+            if (_decorator == null)
+                LOG.warn("Could not create decorator from {}={}", attributeName, decorator);
+            else
+                _context.getObjectFactory().addDecorator(_decorator);
+        }
+    }
+
+    private Decorator asDecorator(Object object)
+    {
+        if (object == null)
+            return null;
+        if (object instanceof Decorator)
+            return (Decorator)object;
+
+        try
+        {
+            Class<?> clazz = object.getClass();
+
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+            final MethodHandle decorate = lookup.findVirtual(clazz, "decorate", decorateType);
+            final MethodHandle destroy = lookup.findVirtual(clazz, "destroy", destroyType);
```

Comment:
Remove the `final` modifiers, unnecessary.

Reply:
No - I'm not going to be in the business of counting lines to keep codestyle happy or not.  We have chosen codestyle to want final in the situation where there is a few lines between declaration and usage, so I'm just using final all the time.  There is no problem with declaring your intention.